python/lsst/ts/scriptqueue/ui/request_model.py

```diff
@@ -0,0 +1,547 @@
+__all__ = ["RequestModel"]
+
+import asyncio
+import logging
+
+from lsst.ts import salobj
+from lsst.ts.scriptqueue import ScriptState, ScriptProcessState
+import SALPY_ScriptQueue
+import SALPY_Script
+
+
+class RequestModel:
+    """A model class that represents the user interface to the queue. It contains stores the current
+    state of the queue as well as for the scripts in the queue. It provides methods to update its
+    internal information and also callback functions that can be scheduled to update in an event loop.
+
+    It also provides methods to send request to the queue and methods to monitor the state of
+    executing script. It relies heavily in the ability of SAL to retrieve late joiner information.
+    """
+
+    def __init__(self, index):
+
+        self.log = logging.getLogger(__name__)
+
+        self.queue = salobj.Remote(SALPY_ScriptQueue, index)
+        self.evt_loop = asyncio.get_event_loop()
+
+        self.cmd_timeout = 120.
+
+        self.state = {'state': "Unknown",
+                      'length': 0,
+                      'past_length': 0,
+                      'queue_scripts': [],
+                      'past_scripts': [],
+                      'current': None}
+        """Store the overall state of the queue.
+        """
+
+        self.scripts = {}
+        """A dictionary to store the state of scripts in the queue. 
+        """
+
+        # Initialize information about the queue. If the queue is not enable or offline state
+        # will be left unchanged.
+        if self.query_queue_state() < 0:
+            self.log.warning("Could not get state of the queue. Make sure queue is running and is "
+                             "in enabled state.")
+            return
+
+        # Initialize information about scripts
+        for i in range(self.state['length']):
+            self.query_script_info(self.state['queue_scripts'][i])
+
+        for i in range(self.state['past_length']):
+            self.query_script_info(self.state['past_scripts'][i])
+
+        if self.state['current'] is not None:
+            self.query_script_info(self.state['current'])
+
+    def enable_queue(self):
```

Comment:
This could be implemented using ts_salobj's new `enable_csc` function (that you are reviewing in DM-17913 https://github.com/lsst-ts/ts_salobj/pull/33/files). Or just ditch this method and call that function directly. It is more robust (I hope) than this code in that it can start from any initial state.

Unfortunately it's clear that with https://jira.lsstcorp.org/browse/DM-18035 it's tricky to determine what state a CSC is actually in. Something I did not take into account when writing that function.